### PR TITLE
Added a callback to show file transfer progress

### PR DESCRIPTION
**Pull request by MohammadAG**
An example callback that prints pushes similar to adb.c:

```
def progress_cb(filename, bytes_written, total_bytes):
    percent = "{0:.2f}".format(100 * (bytes_written / float(total_bytes)))
    print('\r[%s%%] %s' % (percent, filename), end='\r')
    if bytes_written == total_bytes:
        print()
```

### CODE REVIEW

File: adb/filesync_protocol.py [hunk 1]
```diff
@@ -107,11 +109,24 @@ def Push(cls, connection, datafile, filename,
...
+    bytes_written_len = 0
+    total_bytes = -1
+    if progress_callback and isinstance(datafile, file):
+      total_bytes = os.fstat(datafile.fileno()).st_size
+
     while True:
       data = datafile.read(MAX_PUSH_DATA)
-      if not data:
+      if data:
+        if progress_callback:
```
Comment (fahhem) [inline adb/filesync_protocol.py]: you're calling the callback before actually sending data.

Also, can you push the progress-tracking into a small class? That way this loop remains simple and the `callback`/`total_bytes`/`bytes_written_len` is all tracked together

**Comment by MohammadAG** [comment]
Updated according to review and resolved merge conflicts :)

**Comment by MohammadAG** [comment]
Are any other changes needed for this?

File: adb/filesync_protocol.py [hunk 2]
```diff
@@ -107,11 +123,19 @@ def Push(cls, connection, datafile, filename,
     cnxn = FileSyncConnection(connection, b'<2I')
     cnxn.Send(b'SEND', fileinfo)
 
+    if progress_callback:
+      progress = cls._HandlePushProgress(filename, datafile, progress_callback)
```
Comment (fahhem) [inline adb/filesync_protocol.py]: Going with a generator is okay, but you're passing a lot of unnecessary context into it. If you instead passed in a lambda like this:

```py
total_bytes = os.fstat(datafile.fileno()).st_size if isinstance(datafile, file) else -1
progress = cls._HandleProgress(lambda current: progress_callback(filename, current, total_bytes))
```

Now, `_HandleProgress` only tracks things related to the progress: the current bytes and exception ignoring. We could reuse this for anything else that needs progress tracked, like Pull

**Comment by fahhem** [comment]
Sorry about the delay! One last change, but if it's annoying I can merge as-is and make the change myself

**Comment by MohammadAG** [comment]
Done!

**Comment by fahhem** [comment]
Thanks!

**Pull request merged**